### ISSUE TITLE
Orphaned Azure NICs block subnet deletion afterwards

### ISSUE THREAD

**Issue by ialidzhikov** [gardener/machine-controller-manager-provider-azure#41]
**What happened**:
I see quite often failed Azure Shoot deletions with error similar to:

```
-> Pod 'bar.infra.tf-destroy-jv2w9' reported:
* Error deleting Subnet "shoot--foo--bar-nodes" (Virtual Network "shoot--foo--bar" / Resource Group "shoot--foo--bar"): network.SubnetsClient#Delete: Failure sending request: StatusCode=400 -- Original Error: Code="InUseSubnetCannotBeDeleted" Message="Subnet shoot--foo--bar-nodes is in use by /subscriptions/<omitted>/resourceGroups/shoot--foo--bar/providers/Microsoft.Network/networkInterfaces/shoot--foo--bar-cpu-worker-0-z1-7d4cd95c59-p5qgg-nic/ipConfigurations/shoot--foo--bar-cpu-worker-0-z1-7d4cd95c59-p5qgg-nic and cannot be deleted. In order to delete the subnet, delete all the resources within the subnet. See aka.ms/deletesubnet." Details=[]

Releasing state lock. This may take a few moments...
```

Basically the Worker (and machine-controller-manager) resources are deleted but for some reason the NICs are not deleted and block the subnet deletion afterwards (during the Infrastructure deletion).

I can see that the NIC still exists:

```
$ az network nic show -g shoot--foo--bar -n shoot--foo--bar-cpu-worker-0-z1-7d4cd95c59-p5qgg-nic
```

**What you expected to happen**:
No NICs to be orphaned.

**How to reproduce it (as minimally and precisely as possible)**:
Not clear for now.

**Anything else we need to know**
canary # 3637
live # 730
live # 2263
live # 2273

**Environment**:
- machine-controller-manager version: v0.34.3 for buggy `Delete()` API call
- machine-controller-manager version: v0.48.0 for buggy `Get()` API call

**Comment by AxiomSamarth** [comment]
This is now taken care by the Azure Team. Orphan NICs are lately not observed after the changes from Azure on their end. I shall close this issue. The same can be reopened if there are such issues in future again.

**Comment by ialidzhikov** [comment]
/reopen
as we saw ~5 new occurrences of this issue

**Comment by himanshu-kun** [comment]
## Post Grooming discussion + Updates

We reached to Azure and explained that they had a problem with `Get` NIC API call. They responded back with the following
```
We have completed the analysis of your issue with an incorrect Get response being returned for four hours after a resource deletion and recreation.

By default when a resource is deleted, Azure Resource Manager creates a resource consistency job with a delayed start time and a resource long operation job to ensure that the resource is deprovisioned correctly in the control plane caches. For the Network resource provider, they also send a notification to Azure Resource Manager to immediately execute the first delayed resource consistency job when the Network resource provider finishes their asynchronous operation of deprovisioning a deleted resource.
In this case, we believe that a race condition occurred between the background job (resource consistency job) and the Put request which caused the resource to be incorrectly deleted in the Azure Resource Manager cache but successfully provisioned in the Network resource provider.
In most circumstances, the resource would not have shown up again after some time. However, in these circumstances, the resource long operation job from the initial Delete request created a resource consistency job when it completed to make sure the resource was deprovisioned correctly in Azure Resource Manager.
When this resource consistency job executed, it never got a 404 from the Network RP since the resource had been re-provisioned by the Put resource request. There is a 4-hour timeout on these jobs, and when the timeout occurred, the job "re-provisioned" the resource in the ARM cache which caused it to finally appear for the user.
As the problem was caused by a race condition, it is expected to occur infrequently and sporadically. Even though the system will ultimately ensure that the resource state is consistent, it may take up to four hours, as observed in this case. To address this issue, the Azure Resource Manager team will reduce the timeout, which should shorten the time it takes for the resource to return to its expected state. A more permanent solution is currently under consideration.
```

Since we don't have a permanent solution , we have decided NOT to depend on the `Get()` call before issuing a `Delete()` call , but issue `Delete()` directly.
We will still keep `Get()` after the `Delete()` to ensure NIC is gone.

**Comment by unmarshall** [comment]
In case the GetVM calls returns a 404 (not found) but there exists a NIC that as VirtualMachine SubResource still set (non-empty) and if we try and issue a DELETE for the NIC then it will return the following error:
```
RESPONSE 400: 400 Bad Request
ERROR CODE: NicInUse
--------------------------------------------------------------------------------
{
  "error": {
    "code": "NicInUse",
    "message": "Network Interface /subscriptions/82b44c79-a5d4-4d74-8ff8-8639e79c1c39/resourceGroups/shoot--mb-garden--sdktest/providers/Microsoft.Network/networkInterfaces/shoot--mb-garden--sdktest-worker-bingo-nic-alpha is used by existing resource /subscriptions/82b44c79-a5d4-4d74-8ff8-8639e79c1c39/resourceGroups/shoot--mb-garden--sdktest/providers/Microsoft.Compute/virtualMachines/shoot--mb-garden--sdktest-worker-bingo. In order to delete the network interface, it must be dissociated from the resource. To learn more, see aka.ms/deletenic.",
    "details": []
  }
}
```
This will be returned to MCM and a retry will happen. So there is no real need to make a GET call.

**Comment by rishabh-11** [comment]
/close

fixed by #105
